### PR TITLE
Reference for stage 4 symbols-as-weakmap-keys

### LINKED ISSUE #25544: Symbol as WeakMap key is now allowed
### MDN URL

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap

### What specific section or headline is this issue about?

Description

### What information was incorrect, unhelpful, or incomplete?

"[Primitive data types](https://developer.mozilla.org/en-US/docs/Glossary/Primitive) as keys are not allowed (e.g. a [Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) can't be a WeakMap key)."

### What did you expect to see?

WebKit implemented Symbols as keys in WeakMap in https://github.com/WebKit/WebKit/pull/6570

https://github.com/tc39/proposal-symbols-as-weakmap-keys is Stage 3

### Do you have any supporting links, references, or citations?

_No response_

### Do you have anything more you want to share?

The tc39 proposal description says "Permit Symbols as keys in WeakMaps, entries in WeakSets and WeakRefs, and registered in FinalizationRegistries " so there may be other MDN pages that need updating accordingly.

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `en-us/web/javascript/reference/global_objects/weakmap`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/javascript/reference/global_objects/weakmap/index.md
* Last commit: https://github.com/mdn/content/commit/11eb3b9de211e7bc91a31bdc99f268c1c2f93d7b
* Document last modified: 2023-02-21T23:38:32.000Z

</details>

### PR DESCRIPTION
**Pull request by Josh-Cena**
### Description

This adds docs for the [symbol-as-weakmap-keys](https://github.com/tc39/proposal-symbols-as-weakmap-keys) feature. It's implemented in V8 and Safari TP. The FF bug is here: https://bugzilla.mozilla.org/show_bug.cgi?id=1710433

cc @hamishwillee 

### Additional details

This is pending BCD. In addition, I think we should add a note about the limited support for symbol keys at least for a few months, but I don't see a clean way to do it because it's so intertwined with the API's usage.

### Related issues and pull requests

Fix https://github.com/mdn/content/issues/25544

### CODE REVIEW

**Comment by hamishwillee** [comment]
@Josh-Cena I'll leave it to people with more JavaScript expertise to review properly, but you can ping me again Tuesday if you need a slightly less technical review. On quick scan looks great.

FWIW
- I wouldn't push this in until the BCD entry is updated.
- If you do decide on a note about this, then near the top of `files/en-us/web/javascript/reference/global_objects/weakmap/index.md` would be OK
2. What happens if you put in a symbol in a version that only supports objects? What I'm thinking here is that this is what is going to happen, so it might be useful to add some troubleshooting/error information somewhere - at least then the error might be found on search and you could cross-link to compat.
3. The proposal still indicates stage 3. How about adding a note to the FF bug noting the change - the priority might change if they are aware of the compatibility issue.

**Comment by Josh-Cena** [comment]
> 2\. What happens if you put in a symbol in a version that only supports objects?

This throws the same error as if you put in any other primitive value. It's currently documented in https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/No_non-null_object but I highly doubt if this is up-to-date.

> The proposal still indicates stage 3.

It just got advanced in the February meeting and the proposal isn't updated yet. However the spec PR could be merged at any time.

> How about adding a note to the FF bug noting the change

My hope is that they are already aware ^^ FWIW this proposal got to stage 4 a bit faster than I expected, since only Chrome is shipping it in stable currently.

File: files/en-us/web/javascript/reference/errors/no_non-null_object/index.md [hunk 1]
```diff
@@ -60,7 +60,7 @@ Object.defineProperty({}, "key", { value: "foo", writable: false });
 
 ### WeakMap and WeakSet objects require object keys
```
Comment (Elchi3) [inline files/en-us/web/javascript/reference/errors/no_non-null_object/index.md:61]:
```suggestion
### WeakMap and WeakSet objects require object or symbol keys
```

**Review by Elchi3 — APPROVED** [review]
This is great work! I see this is shipping in Safari 16.4 as well, so now there are two stable implementations and we don't need to mark it experimental. I guess we could add a BCD row given it is a change that came a lot later than the initial WeakMap/WeakSet implementations.

I think the important technical piece here is that registered symbols are not part of the deal and you have mentioned that nicely throughout. I'm r+!

**Pull request merged**